### ISSUE TITLE
Add implementations of composite specifications

### ISSUE THREAD

**Issue by MisinformedDNA** [ardalis/Specification#122]
Should composite specifications like `AddSpecification` and `OrSpecification` be added to the library?

**Comment by ardalis** [comment]
Folks can build such things themselves, but I'm not generally a fan of them. The issue is that they start to move logic out of the domain and the specifications and into the application or UI layer where the specifications are created. Now instead of just newing up the appropriate spec and passing it to the repository, the client code needs to stitch together some collection of specs to be applied using boolean logic in a non-standard manner (i.e. not using && and II for boolean logic). All of that composition of filters could have and I would say should have been done in a new specification that simply does that work, has a name, and lives in the domain model.

That said if you need it, and you understand the potential consequences but are fine with them, build it and have at it.

**Comment by MisinformedDNA** [comment]
Right, so if I want to create a new spec based on two other specs and keep things DRY, internally I would want the new spec to simply combine the other two instead of having to copy/paste the logic from the specs into a new one.

**Comment by fiseni** [comment]
Composite specifications address a particular problem in the specification pattern, but are not applicable for Query Specifications.

Specification pattern in general provides a mechanism for combining and encapsulating a set of business rules. The idea is that the given specification will expose `IsValid` behavior which provides information if the defined business rules are met or not. Simply said, they're one form of validators. In this context, creating an infrastructure for composite specifications is quite straightforward. All you have to do is apply some boolean operation to a collection of specifications.

This library has a more specific application. We're trying to leverage the original idea, but instead of representing business rules, in our case specifications describe and represent queries. That said, combining specifications is not feasible, nor logical. Imagine if in one specification you have applied `Take` and `Skip`, and in the second one, you have an additional `Skip`. What `And`/`Or` operations will represent in this case? Yes, composite specifications are feasible even for query specifications, but only for simple implementations where you have only filtering logic (i.e. `Where` expressions possibly).

Makes sense?

**Comment by MisinformedDNA** [comment]
> DRY is a nice and very useful pattern. But, as with everything else, we should not take it to an extreme.

No offense, but that's funny to me. I feel like limiting the specification pattern based on your personal preference as opposed to the guidelines that Martin Fowler set out and which all(/many?) other implementations follow is a bit extreme. Instead of letting users decide how they want to structure their specifications, you are deciding for them.

Combining specs by creating new ones is only trivial in simple scenarios. What if I don't know all the specs that I need upfront? What if a library defines a spec and I want to use it in combination with my own. Am I supposed to decompile their code and copy it?

This greatly increases the learning curve and the amount of work to get started. Not having composites is also contrary to what developers expect leading to confusion (which I've had a lot of).

> Imagine if in one specification you have applied `Take` and `Skip`, and in the second one, you have an additional `Skip`. What `And`/`Or` operations will represent in this case?

OK, so what if there was an interface that didn't allow `Take` and `Skip`, something like `IQueryableSpecification`?



Finally, are composite specs feasible or not in this library, I feel like I got different answers from each of you.

**Comment by ardalis** [comment]
They're feasible (as in possible), but neither Fati nor I recommend them, so we're not including them in this library for the reasons we've tried to outline. This isn't the only library available, and you can always fork it, or just take all the code and use it however you like (it's MIT licensed). But since we don't think they're something most users would benefit from, we've chosen not to include such composite specs (AndSpec, OrSpec, etc.) "in the box".

**Comment by fiseni** [comment]
I'm not offended, enjoy 😄 It's all part of the discussion.

Yes, I'm very well aware of Martin Fowler's guidelines. But, as mentioned above we're not implementing exactly that. If analyzed carefully, Martin talks about creating a mechanism for better encapsulation of the business criteria in your application. There was an initiative to add validation behavior in this library, I was elaborating that in more detail [here](https://github.com/ardalis/Specification/issues/111#issuecomment-803461749).

In our case, the specification should be independent constructs that are closely related to your application, and your application only. They're tightly related to your entities, and inheriting from external sources does not make much sense.

I provided the `Take` and `Skipt` just as a sample, but that applies to all the features we have here, starting from ordering, caching, ORM-specific features like AsNoTracking, etc. Also, if we have noticed we're keeping all criteria as Expressions, not delegates since we're trying to delay the execution and leave that part to ORM providers. So, not even the `Where` feature is not easily composable. We're compiling them for the in-memory evaluations only.

Simply said, in our case specifications are just constructs that describe a given query. They do not act upon the information they hold, they're just information providers to anyone who wants to utilize it (i.e. EF Core evaluators).

Steve's and my answers are aligned actually.
